### DISCUSSION TITLE
Testing rappid with Jest/Enzyme

### DISCUSSION THREAD

**Discussion by Anuthama-Ganesan** [clientIO/joint#1842] [Q&A]
This question is with reference to - https://github.com/clientIO/joint/issues/1475

I am trying to add a test with jest & enzyme for components using rappid.js. And I am encountering the same error TypeError: Cannot read property 'prototype' of undefined.
I have added the mock statement - window.SVGPathElement = jest.fn(); and also added the code mentioned in setupTests.js as per the commit.
node: v16.15.0
jest: 27.5.2,
Is there anything that I am missing?

**Comment by jamesgeorgewilliams** [comment]
It is difficult for us to diagnose the issue without having the code. If you could provide a repository(rappid.js package code not included), or sample code online that would help us.

More generally, `Jest` uses `JSDOM` which doesn't support a lot of SVG APIs.
https://github.com/jsdom/jsdom/issues/918 (If you search the JSDOM github account, you will find other similar discussions)

https://github.com/clientIO/joint-plus-tutorial-react
We have fixed some of the issues in this repository with the fixes you referenced above. Currently the tests are working fine here (This repository is using create-react-app).

It is possible you will meet other functionality that needs to be mocked in Jest.

**Reply by Anuthama-Ganesan** [reply]
I am adding some details here -

This is the error that I encounter:
```
TypeError: Cannot read properties of undefined (reading 'prototype')

      2 | import { useDispatch, useSelector } from 'react-redux';
      3 | import { useNavigate } from 'react-router-dom';
    > 4 | import { ui, dia, shapes } from '@clientio/rappid';
        | ^
      5 | import '@clientio/rappid/rappid.css';
      6 | import '@clientio/rappid/plugins/ui/PaperScroller/css/layout.css';
      7 | import './Genealogy.css';

      at node_modules/@clientio/rappid/rappid.js:4070:32
      at factory (node_modules/@clientio/rappid/rappid.js:3763:3)
      at node_modules/@clientio/rappid/rappid.js:2:65
      at Object.<anonymous> (node_modules/@clientio/rappid/rappid.js:1:2)
      at Object.<anonymous> (node_modules/@clientio/rappid/index.ts:1:1)
      at Object.<anonymous> (src/Components/Genealogy/Genealogy.tsx:4:1)
      at Object.<anonymous> (src/__tests__/Components/Genealogy.test.tsx:10:1)
      at TestScheduler.scheduleTests (node_modules/@jest/core/build/TestScheduler.js:333:13)
      at runJest (node_modules/@jest/core/build/runJest.js:404:19)
      at _run10000 (node_modules/@jest/core/build/cli/index.js:320:7)
      at runCLI (node_modules/@jest/core/build/cli/index.js:173:3)
```
This is my test file 
```
import SelectedUnitRow from '../../Components/SelectedUnitRow';
import React from 'react';
import ReactDOM from 'react-dom';
import { configure, shallow, mount } from "enzyme";
import Adapter from "enzyme-adapter-react-16";
import { Provider } from 'react-redux';
import configureStore from '../../redux/store';
import { act } from "react-dom/test-utils";
import Genealogy from '../../Components/Genealogy/Genealogy'
import '../../setupTests'

configure({ adapter: new Adapter() })
let component;
let parentComponent;
let store;
const configureStoreWrapper = (initialState = {}) => {
    return configureStore(history);
};

window.SVGPathElement = jest.fn();

const setUp = (initialState = {}, isEdit) => {
    store = configureStoreWrapper(initialState);
    const wrapper = mount(
        <Provider store={store}>
            <Genealogy />
        </Provider>
    );
}

describe('Genealogy screen', () => {

    beforeEach(() => {
        component = setUp({}, true);
        Object.defineProperty(global.SVGSVGElement.prototype, 'createSVGMatrix', {
            writable: true,
            value: jest.fn().mockImplementation(() => ({
                martix: jest.fn(() => [[]]),
                a: 0,
                b: 0,
                c: 0,
                d: 0,
                e: 0,
                f: 0,
                flipX: jest.fn().mockImplementation(() => global.SVGSVGElement),
                flipY: jest.fn().mockImplementation(() => global.SVGSVGElement),
                inverse: jest.fn().mockImplementation(() => global.SVGSVGElement),
                multiply: jest.fn().mockImplementation(() => global.SVGSVGElement),
                rotate: jest.fn().mockImplementation(() => ({
                    translate: jest.fn().mockImplementation(() => ({
                        rotate: jest.fn(),
                    })),
                })),
                rotateFromVector: jest.fn().mockImplementation(() => global.SVGSVGElement),
                scale: jest.fn().mockImplementation(() => global.SVGSVGElement),
                scaleNonUniform: jest.fn().mockImplementation(() => global.SVGSVGElement),
                skewX: jest.fn().mockImplementation(() => global.SVGSVGElement),
                skewY: jest.fn().mockImplementation(() => global.SVGSVGElement),
                translate: jest.fn().mockImplementation(() => ({
                    multiply: jest.fn().mockImplementation(() => ({
                        multiply: jest.fn().mockImplementation(() => global.SVGSVGElement),
                    })),
                })),
            })),
        });

        Object.defineProperty(global.SVGSVGElement.prototype, 'createSVGPoint', {
            writable: true,
            value: jest.fn().mockImplementation(() => ({
                x: 0,
                y: 0,
                matrixTransform: jest.fn().mockImplementation(() => ({
                    x: 0,
                    y: 0,
                })),
            })),
        });

        Object.defineProperty(global.SVGSVGElement.prototype, 'createSVGTransform', {
            writable: true,
            value: jest.fn().mockImplementation(() => ({
                angle: 0,
                matrix: {
                    a: 1,
                    b: 0,
                    c: 0,
                    d: 1,
                    e: 0,
                    f: 0,
                    multiply: jest.fn(),
                },
                setMatrix: jest.fn(),
                setTranslate: jest.fn(),
            })),
        });

    });

    const initialState = {};
    const store = configureStore(initialState);

    it("render Genealogy Component without crash", async () => {
        const div = document.createElement("div");
        await act(async () => {
            ReactDOM.render(component, div);
        });
    });

    it('renders SelectedUnitRow', async () => {
        const div = document.createElement('div');
        await act(async () => {
            ReactDOM.render(component, div);
        });
        expect(component.find(SelectedUnitRow)).to.have.length(1);
    });
});
```
I was aware that `Jest` does not support SVG files and some issues were fixed on the same lines with reference to the issue I mentioned above. I have used those fixes in my file, but still I am running into the issue. The repository is created using create-react-app. Let me know if any other information is required?

**Reply by jamesgeorgewilliams** [reply]
Thanks for the details!

If I understand correctly you are using a `setupTests.js` file, but in the code example above, you seem to have a mixture of tests and the logic that should be in the `setupTests` file.
https://github.com/clientIO/joint-plus-tutorial-react/blob/main/src/setupTests.ts

As far as I am aware the global setup for tests in create-react-app should be placed in `src/setupTests.ts`.
https://create-react-app.dev/docs/running-tests/#initializing-test-environment

**Reply by Anuthama-Ganesan** [reply]
Using it from the `setUpTests.js` file did not seem to do the trick. The `setUpTests` file was placed in the `src` folder.  So, I tried to place the lines of code in the test file itself and ran the tests. either ways, I am running into the same issue. 
Any hunch on what I could be missing?

**Reply by jamesgeorgewilliams** [reply]
Is it possible you can provide your create-react-app code(jointjs+ package not included) publicly on one of the following services?
 https://jsfiddle.net
 https://codesandbox.io
 https://codepen.io

or make it available via a github repo?

**Reply by Anuthama-Ganesan** [reply]
I have added the component file, test file, js test setup file, package.json and other related files in the following link -
https://codesandbox.io/s/hungry-nash-784w8b?file=/src/App.js
Let me know if I have to share any other details. 
Thanks.

**Reply by Anuthama-Ganesan** [reply]
Hey @jamesgeorgewilliams, any updates on this?

**Reply by jamesgeorgewilliams** [reply]
Hi, I wasn't able to run your project as there are a few issues with it.

1) It is not possible to `npm install` in your project. Your dependencies in `package.json` are not compatible with each other. `React` version `18` and `enzyme-adapter-react-16` for example.
2) The project is missing many files. The `Genealogy` component imports other components that weren't provided.
3) There are no tests provided in `Genealogy.test.ts` as the file is empty. 

If you can provide a working project, with all the files completed and provided, that would be great.

**Reply by Anuthama-Ganesan** [reply]
Hi @jamesgeorgewilliams I was able to resolve this issue. Turns out the setUpTests portion was not getting set up properly coz this line was missing in jest.config.json -   setupFilesAfterEnv: ["./src/setupTests.js"]. This resolves the issue.
